alacritty_terminal/src/term/mod.rs

```diff
@@ -1421,17 +1421,32 @@ impl<T: EventListener> Handler for Term<T> {
             ansi::ClearMode::All => {
                 if self.mode.contains(TermMode::ALT_SCREEN) {
                     self.grid.reset_region(..);
-                } else {
+                } else if self.mode.contains(TermMode::VI) {
+                    let old_offset = self.grid.display_offset();
+
                     self.grid.clear_viewport();
 
-                    self.vi_mode_cursor.point.line =
-                        self.vi_mode_cursor.point.line.grid_clamp(self, Boundary::Cursor);
+                    // How many lines has been scrolled by clearing viewport.
+                    let lines = self.grid.display_offset().saturating_sub(old_offset);
+                    self.vi_mode_cursor.point.line.0 -= lines as i32;
+
+                    // Clamp the vi cursor into the display area.
+                    let display_topmost = -(self.grid.display_offset() as i32);
+                    let display_bottommost = display_topmost + (self.screen_lines() as i32) - 1;
+                    self.vi_mode_cursor.point.line = Line(max(
+                        min(self.vi_mode_cursor.point.line.0, display_bottommost),
+                        display_topmost,
+                    ));
+                } else {
+                    self.grid.clear_viewport();
+                    self.grid.reset_display_offset();
```

Comment:
Why do we reset the display offset conditionally? That just seems inconsistent. Either we always go back to the bottom, or we never do. The vi cursor's presence should have no impact on this.

Reply:
The reason is to keep Alacritty's previous one on vi-mode when viewport clear is invoked. I'm interested in the problem on vi-mode, and I don't know an appropriate behavior on not vi-mode when viewport clear is invoked. For example, Xterm, which is a defacto standard terminal emulator, and urxvt reset display offset when that is done, however many other terminal emulators don't do that. I may prefer the latter's style if anything. Additionally, current Alacritty changes display offset for every new output to the viewport. The latter's style may be consistent with that.

Reply:
Most terminals probably just scroll down if there's any "write" to the viewport, including clears (XTerm and URxvt certainly do). Which isn't something Alacritty does and I generally disagree with the behavior.

I'm not sure I care particularly much either way. But it **must** be consistent regardless of vi mode. So if we scroll down without it, we need to scroll down with it. And to be consistent with our other logic (i.e. not scrolling down on new output), I think keeping the offset makes sense.

Reply:
> I'm not sure I care particularly much either way. But it must be consistent regardless of vi mode. So if we scroll down without it, we need to scroll down with it. And to be consistent with our other logic (i.e. not scrolling down on new output), I think keeping the offset makes sense.

I got it. I'll fix to keep the display offset whether or not vi mode is active. Thanks for your comment!